---
title: Visualizing allele effects patterns across traits that map to chromosome 2
  hotspot
author: "Frederick Boehm"
date: "9/10/2018"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal

Goal in this Rmd file is to write code needed for creation of a heatmap of allele effects values for every trait that maps to chromosome 2 hotspot.

## Resources for heatmaps in R

Here are links to some potentially valuable resources for heatmap creation in R:



## Karl's suggestions for clusterings of traits

In our meeting last week (September 6 2018), Karl suggested several ways to cluster or organize the traits within the heatmap. These include:


## Plan

First, we need to scan chromosome 2 with each trait. From the scan outputs, we can calculate the allele effects at the position of the LOD peak. Once we have the 8 founder allele effects per trait, we can create a heatmap

## Univariate scans of chromosome 2

First, we need to read the .Rdata file that we downloaded from DataDryad.

```{r}
load("../data/Attie_DO378_eQTL_viewer_v1.Rdata")
readRDS("../data-to-condor/hotspot_expr_tib2_keller_chr2.rds") -> hotspot_expr
readRDS("../data-to-condor/keller2018-chr2-local-expr.rds") -> local_expr
```

We also loaded the two RDS files that contain the expression values for the traits of interest.

The genotypes probabilities are their own object (loaded with the DataDryad file), `genoprobs`.

```{r}
library(qtl2)
# convert to needed phenotypes input format for scan1
as.matrix(local_expr[, 1:21]) -> local_expr_mat
rownames(local_expr_mat) <- local_expr[, 22]
#
as.matrix(hotspot_expr[, 1:139]) -> hotspot_expr_mat
rownames(hotspot_expr_mat) <- hotspot_expr[, 140]
#
# isolate additive covariates
dataset.islet.rnaseq$covar -> addcovar
```



```{r}
scan1(genoprobs = genoprobs, pheno = local_expr_mat, kinship = K, addcovar = addcovar, reml = TRUE
        ) -> s1_local
scan1(genoprobs = genoprobs, pheno = hotspot_expr_mat, kinship = K, addcovar = addcovar, reml = TRUE
        ) -> s1_hotspot

```

We now search the outputs of `scan1` to identify peaks:

```{r}
find_peaks(s1_local, map = map, threshold = 20) -> local_peaks
find_peaks(s1_hotspot, map = map) -> hotspot_peaks
```

Note above that we use the default value of `threshold` when calling `find_peaks` for the hotspot traits. This ensures that we consider all 169 traits. we'll subset on chromosome 2 peaks below.

Next, we calculate allele effects for each trait on chromosome 2. We keep separate the local and hotspot traits.

```{r}
# local traits
local_effects <- list()
for (i in 1:21){
  scan1coef(genoprobs = genoprobs[ , 2], 
            pheno = local_expr_mat[ , i],
            kinship = K[[2]],
            addcovar = addcovar,
            reml = TRUE
            ) -> local_effects[[i]]
}    
# hotspot traits
hotspot_effects <- list()
for (i in 1:139){
  scan1coef(genoprobs = genoprobs[ , 2], 
            pheno = hotspot_expr_mat[ , i],
            kinship = K[[2]],
            addcovar = addcovar,
            reml = TRUE
            ) -> hotspot_effects[[i]]
}    
```

Now, we want to isolate, for each trait, the 8-long vector of founder allele effects at the trait's corresponding peak position.

We need to use the R object `map` (from the Datadryad file).

```{r}
map[[2]]
```

Note that in our output from the calls to `find_peaks`, we have position in Mb. However, in the outputs of the calls to `scan1coef`, we have no positions - we do, however, know that the marker ordering is maintained.

So, we need to identify the index - on chromosome 2 - that corresponds to each trait's peak position.

```{r}
library(tidyverse)
# convert map for chr2 to a tibble
m2 <- map[[2]] %>% 
  as_tibble() %>%
  mutate(index = 1:4709)
local_peaks2 <- local_peaks %>%
  as_tibble() %>%
  left_join(m2, by = c("pos" = "value"))
hotspot_peaks2 <- hotspot_peaks %>%
  as_tibble() %>%
  filter(chr == "2") %>%
  left_join(m2, by = c("pos" = "value"))
```

We define a helper function:

```{r}
get_effects <- function(marker_index, allele_effects_matrix, map){
  marker_name <- names(map[marker_index])
  allele_effects_matrix[rownames(allele_effects_matrix) == marker_name, 1:8]
}
```

```{r}
hotspot_effects_at_peaks <- list()
for (i in 1:nrow(hotspot_peaks2)){
  hotspot_effects_at_peaks[[i]] <- get_effects(hotspot_peaks2$index[i], allele_effects_matrix = hotspot_effects[[i]], map = map[[2]]) # need map for chromosome 2 here.
}
local_effects_at_peaks <- list()
for (i in 1:nrow(local_peaks2)){
  local_effects_at_peaks[[i]] <- get_effects(local_peaks2$index[i], allele_effects_matrix = local_effects[[i]], map = map[[2]]) # need map for chromosome 2 here.
}
```

We then arrange the two sets of allele effects into matrices.

```{r}
# organize effects into a matrix
local_effects_at_peaks_mat <- sapply(FUN = function(x)x, X = local_effects_at_peaks) %>% t()
rownames(local_effects_at_peaks_mat) <- local_peaks2$lodcolumn
hotspot_effects_at_peaks_mat <- sapply(FUN = function(x)x, X = hotspot_effects_at_peaks) %>% t()
rownames(hotspot_effects_at_peaks_mat) <- hotspot_peaks2$lodcolumn
```


## R code for heatmaps

I now want to visualize the two matrices of allele effects with heatmaps.

To use the ggplot2 package we need to first arrange the data in tidy format. One row per founder allele value, with additional columns to specify which allele and which gene. So it will be a 1112 x 3 tibble.

We want to be careful when calling a function such as `as.vector` to arrange the values in the matrix.

```{r}
foo <- matrix(1:8, nrow = 2)
foo
as.vector(foo)
bar <- matrix(1:8, nrow = 2, byrow = TRUE)
bar
as.vector(bar)
```

We see from the above that `as.vector` concatenates columns of a matrix.

We first prepare a helper function:

```{r}
tidy_effects <- function(effects_mat){
  effect_size <- as.vector(effects_mat)
  founder_allele <- rep(colnames(effects_mat), each = nrow(effects_mat))
  gene_expr_id <- rep(rownames(effects_mat), times = ncol(effects_mat))
  return(tibble(gene_expr_id, founder_allele, effect_size))
}
```


Now, we input an allele effects matrix to `tidy_effects`:

```{r}
tidy_effects(local_effects_at_peaks_mat) %>%
  ggplot(aes(y = gene_expr_id, x = founder_allele)) +
  geom_tile(aes(fill = effect_size)) 
```

## Clustering genes' allele effects

```{r}
local_effects_at_peaks_mat %>%
  t() %>%
  cor() %>%
  is.na() %>%
  sum()
local_effects_at_peaks_mat %>%
  #t() %>%
  dist() %>%
  hclust() %>%
  plot()
local_hc <- local_effects_at_peaks_mat %>%
  dist() %>%
  hclust()
```

```{r}
tidy_effects(local_effects_at_peaks_mat[local_hc$order, ]) %>%
  ggplot(aes(y = gene_expr_id, x = founder_allele)) +
  geom_tile(aes(fill = effect_size)) 
```

Heatmap above has rows ordered in numerical order. Yet, I called the function `tidy_effects`.

## Heatmaps for nonlocal traits that map to chromosome 2 hotspot

```{r}
tidy_effects(hotspot_effects_at_peaks_mat) %>%
  ggplot(aes(y = gene_expr_id, x = founder_allele)) +
  geom_tile(aes(fill = effect_size)) 
```

The "E" allele seems to have high values in a large number of traits.

Let's also do hierarchical clustering on the genes' allele effects and reorder them before plotting. 




## Using gene expression data to hierarchically cluster (and order) genes

Rather than using allele effects to cluster genes, let's look use the raw expression values.

```{r}
hotspot_effects_at_peaks_mat %>%
  dist() %>%
  hclust() %>%
  plot()
hotspot_hc <- hotspot_effects_at_peaks_mat %>%
  dist() %>%
  hclust()
```

```{r}
tidy_effects(hotspot_effects_at_peaks_mat[hotspot_hc$order, ]) %>%
  ggplot(aes(y = gene_expr_id, x = founder_allele)) +
  geom_tile(aes(fill = effect_size)) 
```

## Order nonlocal traits by value of "E" allele

```{r}
hotspot_effects_at_peaks_mat[order(hotspot_effects_at_peaks_mat[ , 8]), ] %>%
  tidy_effects() %>%
  ggplot(aes(y = gene_expr_id, x = founder_allele)) +
  geom_tile(aes(fill = effect_size)) 
```


## Order nonlocal traits by value of "H" allele

My initial attempt at this failed, as I was unaware of hte fact that geom_tile function in ggplot2 does its own sorting by name.



## Trying to understand ordering in plots when using `geom_tile`

It seems that `geom_tile` applies an ordering step when deciding ordering of rows and columns in my attempts at heatmaps. Is there another function that I might use? Base R has `heatmap`.


